cmd/nerdctl/logs.go

```diff
@@ -106,83 +113,45 @@ func logsAction(cmd *cobra.Command, args []string) error {
 			if err != nil {
 				return err
 			}
-			logConfigFilePath := logging.LogConfigFilePath(dataStore, l[labels.Namespace], found.Container.ID())
-			var logConfig logging.LogConfig
-			logConfigFileB, err := os.ReadFile(logConfigFilePath)
+			task, err := found.Container.Task(ctx, nil)
 			if err != nil {
 				return err
 			}
-			if err = json.Unmarshal(logConfigFileB, &logConfig); err != nil {
+			status, err := task.Status(ctx)
+			if err != nil {
 				return err
 			}
-			switch logConfig.Driver {
-			// TODO: move these logics to pkg/logging
-			case "json-file":
-				logJSONFilePath := jsonfile.Path(dataStore, ns, found.Container.ID())
-				if _, err := os.Stat(logJSONFilePath); err != nil {
-					return fmt.Errorf("failed to open %q, container is not created with `nerdctl run -d`?: %w", logJSONFilePath, err)
-				}
-				task, err := found.Container.Task(ctx, nil)
-				if err != nil {
-					return err
-				}
-				status, err := task.Status(ctx)
-				if err != nil {
-					return err
-				}
-				if status.Status != containerd.Running {
-					follow = false
-				}
+			if status.Status != containerd.Running {
+				follow = false
+			}
 
-				reader, execCmd, err := newTailReader(ctx, logJSONFilePath, follow, tail)
-				if err != nil {
-					return err
-				}
-				go func() {
-					if follow {
-						if waitCh, err := task.Wait(ctx); err == nil {
-							<-waitCh
-						}
-						execCmd.Process.Kill()
-					}
-				}()
-				return jsonfile.Decode(os.Stdout, os.Stderr, reader, timestamps, since, until)
-			case "journald":
-				shortID := found.Container.ID()[:12]
-				var journalctlArgs = []string{fmt.Sprintf("SYSLOG_IDENTIFIER=%s", shortID), "--output=cat"}
+			// Setup goroutine to send stop event if container task finishes:
+			go func() {
 				if follow {
-					journalctlArgs = append(journalctlArgs, "-f")
-				}
-				if since != "" {
-					// using GetTimestamp from moby to keep time format consistency
-					ts, err := timetypes.GetTimestamp(since, time.Now())
-					if err != nil {
-						return fmt.Errorf("invalid value for \"since\": %w", err)
-					}
-					date, err := prepareJournalCtlDate(ts)
-					if err != nil {
-						return err
-					}
-					journalctlArgs = append(journalctlArgs, "--since", date)
-				}
-				if timestamps {
-					logrus.Warnf("unsupported timestamps option for jounrald driver")
-				}
-				if until != "" {
-					// using GetTimestamp from moby to keep time format consistency
-					ts, err := timetypes.GetTimestamp(until, time.Now())
-					if err != nil {
-						return fmt.Errorf("invalid value for \"until\": %w", err)
-					}
-					date, err := prepareJournalCtlDate(ts)
-					if err != nil {
-						return err
+					if waitCh, err := task.Wait(ctx); err == nil {
+						<-waitCh
+						logrus.Debugf("container task has finished, sending kill event to log viewer")
+						stopChannel <- os.Interrupt
```

Comment:
Great catch, seeing as though a [restart is just a sequential stop and start](https://github.com/containerd/nerdctl/blob/master/cmd/nerdctl/restart.go#L66-L71), this would kill the existing `task` and thus stop the `logs -f` command too. (just tested it myself and can confirm it stops)

We could simply requery the container's current task and status but I'm worried we could catch the container "in between restarts" and before it actually reaches `Up` status, but I'll try to figure something out...

Reply:
@fahedouch I've made some [changes for working around a container restart](https://github.com/aznashwan/nerdctl/commit/2c980ea72cd76e2f0bf76abb82598d894b28d88c) on a separate branch but unfortunately the JSON file based logging [rotates the logfile](https://github.com/containerd/nerdctl/blob/master/pkg/logging/json_logger.go#L80-L82) by moving `$CONTAINER_ID-json.log` to `$CONTAINER_ID-json.log.1` every time a container is started, which leads to:
* the `tail`-based implementation failing to re-open the new logfile, so `tail` keeps feeding the old `$CONTAINER_ID-json.log.1` file which is no longer being written to until the container is finally stopped or exists
* the native implementation will attempt re-open the file at the original path but preserve the file seek, which could lead to it seeking over the rotated logfile's fill

Although both of the above could be theoretically bypassed by sending a `syscall.SIGHUP` on the `stopChannel` and having the `LogViewer` refresh the log viewing, I don't consider the technical overhead to be worth it for something this edge-casey...

Please note I've also tested `docker logs` from Docker CE 20.10.18 and can report that:
* `docker logs -f` also closes itself if the container `docker restart`ed so in that sense the current behavior is correct
* `docker logs` does **not** seem to rotate logfiles so us using logrotate is deviating a bit from how Docker stores logs

Reply:
Regarding log file rotation, I confirm that after restart nerdctl writes in the principal log file.

benchmark of mod time after a restart of the container  `nerdctl restart  0e8a3fdbf637` : 

```
root@e4276e97ebd3:/home/containerd/nerdctl# stat /var/lib/nerdctl/1935db59/containers/default/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f-json.log
  File: /var/lib/nerdctl/1935db59/containers/default/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f-json.log
  Size: 150083          Blocks: 296        IO Block: 4096   regular file
Device: 10000eh/1048590d        Inode: 3708060     Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-11-01 11:18:47.922219013 +0000
Modify: 2022-11-01 11:17:45.048155011 +0000
Change: 2022-11-01 11:17:45.048155011 +0000
 Birth: 2022-11-01 11:17:45.035155011 +0000
root@e4276e97ebd3:/home/containerd/nerdctl# stat /var/lib/nerdctl/1935db59/containers/default/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f-json.log.1 
  File: /var/lib/nerdctl/1935db59/containers/default/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f/0e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f-json.log.1
  Size: 150128          Blocks: 296        IO Block: 4096   regular file
Device: 10000eh/1048590d        Inode: 3708071     Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-11-01 11:17:45.035552345 +0000
Modify: 2022-11-01 11:17:45.035552470 +0000
Change: 2022-11-01 11:17:45.034155011 +0000
 Birth: 2022-11-01 11:12:35.079202007 +0000
````

mtime of `e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f-json.log` > mtime of `e8a3fdbf637c61db53536d9ce4a6334726b432c5e1b6101fd71e9c48d7dc93f-json.log.1`

Reply:
> I think we can make it simpler by checking wether the container exists or not
>
> The principal action(s) that can be exerted on the container that we follow its logs are:
> 
> nerdctl stop => do not trigger the stopChannel
> nerdctl stop => nerdctl start => do not trigger the stopChannel
> nerdctl restart => do not trigger stopChannel
> nerdctl kill => do not trigger the stopChannel, the container exists by no longer emitting logs
> nerdctl stop => nerdctl rm => do trigger stopChannel
> ==> we Interrupt only when the container is removed

If we were to solely rely on the container's existence, it would make `nerdctl logs -f` on a running container never stop even if the container stops. (neither by it stopping by itself or being stopped by another command)

As mentioned, `docker logs -f` exits if the container stops, and there's also a check which looks like it was meant to [explicitly test that `logs -f` exits when a container does](https://github.com/containerd/nerdctl/blob/master/cmd/nerdctl/logs_test.go#L136-L137), so it looks we'll need to check the container's running status one way or the other.

> @fahedouch apologies for the confusion, I wasn't saying nerdctl is logging to the wrong file, I was saying that the tail executable will continue reading from the old file (which has been renamed to *-json.log.1) instead of the new file (*-json.log).
>
> Although none of the above are that involved, I honestly believe it'd add more code overhead than it's worth for such an edge-casey situation, but please let me know which approach you think is best.

The question of whether or not these workarounds are worth the risk of a blocking/failing `nerdctl logs -f` command still stands. (IMHO it adds too much overhead for a behaviour `docker logs -f` itself doesn't do)

Reply:
> If we were to solely rely on the container's existence, it would make nerdctl logs -f on a running container never stop even if the container stops. (neither by it stopping by itself or being stopped by another command)
As mentioned, docker logs -f exits if the container stops, and there's also a check which looks like it was meant to [explicitly test that logs -f exits when a container does](https://github.com/containerd/nerdctl/blob/master/cmd/nerdctl/logs_test.go#L136-L137), so it looks we'll need to check the container's running status one way or the other.

@aznashwan Agree apologies for the confusion. Yes to be compatible with docker we should consider that following logs is conditioned by container state. so I am ok with this [code](https://github.com/aznashwan/nerdctl/commit/2c980ea72cd76e2f0bf76abb82598d894b28d88c).

I got your point regarding tail. We have delegated the read responsibility to tail. It looks like it has a limitation when the file inode number changes. I believe we can put it aside for investigation in another PR may be. let's keep the actual behavior for the moment

Reply:
> Yes to be compatible with docker we should consider that following logs is conditioned by container state. so I am ok with this [code](https://github.com/aznashwan/nerdctl/commit/2c980ea72cd76e2f0bf76abb82598d894b28d88c).

@fahedouch unfortunately that was the commit which initially lead me to discover [the issues caused by the log file being rotated](https://github.com/containerd/nerdctl/pull/1468#discussion_r1010483455) with both `tail` and the native implementation.

> I believe we can put it aside for investigation in another PR may be. let's keep the actual behavior for the moment :)

In this sense, I have ["downgraded" the logic](https://github.com/containerd/nerdctl/pull/1468/files#diff-7fd567884e5d5e48e4d2b715637bc3d446f402579041620814c0da24cd9c271dR136-R138) to immediately send the stop signal when the current container task finishes [as is the case in the current main branch](https://github.com/containerd/nerdctl/blob/master/cmd/nerdctl/logs.go#L144-L146).